src/viam/ur/module/ur_arm.cpp

```diff
@@ -172,128 +184,1284 @@ void write_waypoints_to_csv(const std::string& filepath, const std::list<Eigen::
+            const auto current_state_description = current_state.describe();
+            const auto event_description = event.describe();
+            auto new_state = current_state.handle_event(std::forward<T>(event));
+            const auto new_state_description = std::visit([](const auto& state) { return std::string{state.describe()}; }, new_state);
+            VIAM_SDK_LOG(info) << "URArm state transition from state `" << current_state_description << "` to state `"
+                               << new_state_description << "` due to event `" << event_description << "`";
+            return new_state;
+        },
+        current_state_);
+}
+
+std::chrono::milliseconds URArm::state_::get_timeout_() const {
+    return std::visit([](auto& state) { return state.get_timeout(); }, current_state_);
+}
+
+std::string URArm::state_::describe_() const {
+    return std::visit([](auto& state) { return state.describe(); }, current_state_);
+}
+
+void URArm::state_::upgrade_downgrade_() {
+    if (auto event = std::visit([this](auto& state) { return state.upgrade_downgrade(*this); }, current_state_)) {
+        emit_event_(*std::move(event));
+    }
+}
+
+void URArm::state_::clear_ephemeral_values_() {
+    ephemeral_.reset();
+}
+
+void URArm::state_::recv_arm_data_() {
+    if (auto event = std::visit([this](auto& state) { return state.recv_arm_data(*this); }, current_state_)) {
+        emit_event_(*std::move(event));
+    }
+}
 
-    // variables derived from arm state on configuration
-    bool is_sim;
+void URArm::state_::handle_move_request_() {
+    if (auto event = std::visit([this](auto& state) { return state.handle_move_request(*this); }, current_state_)) {
+        emit_event_(*std::move(event));
+    }
+}
+
+void URArm::state_::send_noop_() {
+    if (auto event = std::visit([](auto& state) { return state.send_noop(); }, current_state_)) {
+        emit_event_(*std::move(event));
+    }
+}
+
+void URArm::state_::run_() {
+    VIAM_SDK_LOG(info) << "worker thread started";
+    while (true) {
+        std::unique_lock lock(mutex_);
+        if (worker_wakeup_cv_.wait_for(lock, get_timeout_(), [this] { return shutdown_requested_; })) {
+            VIAM_SDK_LOG(info) << "worker thread signaled to terminate";
+            break;
+        }
+
+        try {
+            clear_ephemeral_values_();
+            recv_arm_data_();
+            upgrade_downgrade_();
+            handle_move_request_();
+            send_noop_();
+        } catch (const std::exception& ex) {
+            VIAM_SDK_LOG(warn) << "Exception in worker thread: " << ex.what();
+        } catch (...) {
+            VIAM_SDK_LOG(warn) << "Unknown exception in worker thread";
+        }
+    }
+
+    VIAM_SDK_LOG(info) << "worker thread emitting disconnection event";
+    emit_event_(event_connection_lost_{});
+    VIAM_SDK_LOG(info) << "worker thread terminating";
+}
+
+void URArm::state_::trajectory_done_callback_(const control::TrajectoryResult trajectory_result) {
```

Comment:
Not much has changed here except that we can now swap the move request out under the lock but run the completion handlers without holding the lock, allowing the worker thread to get back to its job of communicating with the arm.